File: 00978. Longest Turbulent Subarray/solution.js
/**
 * Problem: 978. Longest Turbulent Subarray
 *
 * Difficulty: Medium
 *
 * Language: JavaScript
 *
 * Performance: Runtime - 0 ms (Beats 100%)
 */

/**
 * Finds the length of the longest turbulent subarray
 *
 * @param {number[]} arr - Input array
 *
 * @returns {number} - Maximum turbulent subarray length
 */
const maxTurbulenceSize = (arr) => {
  // Handle the base case of a single element array
  if (arr.length === 1) return 1

  // Initialize tracking variables for increasing and decreasing patterns
  let increasingLen = 1,
    decreasingLen = 1,
    maxLength = 1

  for (let i = 1; i < arr.length; i++) {
    if (arr[i] > arr[i - 1]) {
      // Current element is greater than previous, extend decreasing pattern
      increasingLen = decreasingLen + 1
      decreasingLen = 1
    } else if (arr[i] < arr[i - 1]) {
      // Current element is less than previous, extend increasing pattern
      decreasingLen = increasingLen + 1
      increasingLen = 1
    } else {
      // Equal elements break turbulence, reset both patterns
      increasingLen = decreasingLen = 1
    }

    // Update the maximum length found so far
    maxLength = Math.max(maxLength, increasingLen, decreasingLen)
  }

  return maxLength
}

File: 00978. Longest Turbulent Subarray/solution.go
/**
 * Problem: 978. Longest Turbulent Subarray
 *
 * Difficulty: Medium
 *
 * Language: Go
 *
 * Performance: Runtime - 0 ms (Beats 100%)
 */

func maxTurbulenceSize(arr []int) int {
	// Handle edge case of arrays with less than 2 elements
	if len(arr) < 2 {
		return 1
	}

	// Initialize variables for left boundary, maximum length, and previous difference
	start, maxLength, prevDiff := 0, 0, 0

	// Iterate through the array starting from the second element
	for i, num := range arr[1:] {
		// Calculate current difference between consecutive elements
		currDiff := arr[i] - num

		// Reset start position if elements are equal (no turbulence)
		if currDiff == 0 {
			start = i + 1
		} else if prevDiff == 0 || (prevDiff > 0 && currDiff > 0) || (prevDiff < 0 && currDiff < 0) {
			// Reset start if pattern breaks: same signs or first comparison
			start = i
		}

		// Update previous difference for next iteration
		prevDiff = currDiff
		// Update max length found so far
		maxLength = max(maxLength, i+2-start)
	}

	return maxLength
}